providers/gcs/gcs.go

```diff
@@ -226,12 +226,22 @@ func (b *Bucket) Iter(ctx context.Context, dir string, f func(string) error, opt
 
 // Get returns a reader for the given object name.
 func (b *Bucket) Get(ctx context.Context, name string) (io.ReadCloser, error) {
-	return b.bkt.Object(name).NewReader(ctx)
+	r, err := b.bkt.Object(name).NewReader(ctx)
+	if err != nil {
+		return r, err
+	}
+
+	return objstore.ObjectSizerReadCloser{ReadCloser: r, Size: r.Attrs.Size}, nil
 }
 
 // GetRange returns a new range reader for the given object name and range.
 func (b *Bucket) GetRange(ctx context.Context, name string, off, length int64) (io.ReadCloser, error) {
-	return b.bkt.Object(name).NewRangeReader(ctx, off, length)
+	r, err := b.bkt.Object(name).NewRangeReader(ctx, off, length)
+	if err != nil {
+		return r, err
+	}
+
+	return objstore.ObjectSizerReadCloser{ReadCloser: r, Size: r.Attrs.Size}, nil
```

Comment:
Should the size be set to `length` here?

Reply:
This is where that value is set: https://github.com/googleapis/google-cloud-go/blob/main/storage/grpc_client.go#L1126.

According to the code, this is the size of the entire object, even if a range is requested: https://github.com/googleapis/google-cloud-go/blob/main/storage/grpc_client.go#L1109-L1110.

We might want to test this better to be sure we're returning the right value.

Reply:
thanks! adding it for `GetRange` was an after thought, i'll have another look

Reply:
[`Remain()`](https://github.com/googleapis/google-cloud-go/blob/main/storage/reader.go#L270) looks like the right value to use for Range. wdyt?

https://github.com/googleapis/google-cloud-go/blob/main/storage/grpc_client.go#L1151
https://github.com/googleapis/google-cloud-go/blob/main/storage/http_client.go#L1434

Reply:
Yes, this could be the one that we need. We have e2e tests that run in the CI which could be useful for this: https://github.com/thanos-io/objstore/blob/main/testing.go#L83.

Reply:
other providers are returning content length which i am hoping reflects the requested range, I'll try to do a basic sanity check for a couple of providers tomorrow.

if this feels risky, i can limit the change to `Get` call alone

Reply:
👍 i'll adding something there :)

Reply:
@fpetkovski added the tests, it's looking good. can you take another look when you get a chance? Thanks!